src/libraries/System.Net.Security/src/System/Net/CertificateValidationPal.Unix.cs

```diff
@@ -88,14 +83,11 @@ internal static partial class CertificateValidationPal
             }
             finally
             {
-                if (remoteContext != null)
+                if (remoteCertificate != IntPtr.Zero && result == null)
```

Comment:
`new X509Certifiate2(IntPtr)` is going to upref the handle, so your call to free shouldn't be conditional on the instance succeeding.